src/backend/controllers/pages.ts

```diff
@@ -62,14 +64,110 @@ class Pages {
     return nullFilteredPages;
   }
 
+  /**
+   * Group all pages by their parents
+   * If the pageId is passed, it excludes passed page from result pages
+   *
+   * @param {string} pageId - pageId to exclude from result pages
+   * @returns {Page[]}
+   */
+  public static async groupByParent(pageId = ''): Promise<Page[]> {
+    const result: Page[] = [];
+    const orderGroupedByParent: Record<string, string[]> = {};
+    const rootPageOrder = await PagesOrder.getRootPageOrder();
+    const childPageOrder = await PagesOrder.getChildPageOrder();
+    const orphanPageOrder: PageOrder[] = [];
+
+    /**
+     * If there is no root nad child page order, then it returns empty array
+     */
+    if (!rootPageOrder || (!rootPageOrder && childPageOrder.length <= 0)) {
+      return [];
+    }
+
+    const pages = (await this.getAll()).reduce((map, _page) => {
+      map.set(_page._id, _page);
+
+      return map;
+    }, new Map);
+    const rootPagesId = rootPageOrder.order;
+
+    /**
+     * It groups root pages and 1 level pages by its parent
+     */
+    rootPagesId.reduce((prev, curr, idx) => {
+      const childPages = childPageOrder.filter((child, _idx) => {
+        if (child.page === curr) {
+          childPageOrder.splice(_idx, 1);
```

Comment:
I think, it would better to separate the `filter` method from the logic lice splicing the external array. Filter supposed to only filter the incoming array using some statement.


Maybe something like that would be better understandable?

```js
const childPages = childPageOrder
  .filter(pageOrder.page => pageOrder.page === cur)
  .forEach(pageOrder => {
    childPageOrder.splice(...)
  })
````

Reply:
Fixed